src/Npgsql/NpgsqlRawReplicationStream.cs

```diff
@@ -0,0 +1,552 @@
+                            }
+                        }
+                        EndOfStream = true;
+                        throw;
+                    }
+                    
+                    if (msg == null)
+                    {
+                        if (dontWait)
+                            return status;
+                        continue;
+                    }
+                    status = NpgsqlReplicationStreamFetchStatus.None;
+
+                    switch (msg.Code)
+                    {
+                        case BackendMessageCode.CopyData:
+                            _bytesLeft = ((CopyDataMessage)msg).Length;
+                            _length = _bytesLeft;
+                            break;
+                        case BackendMessageCode.CopyDone:
+                            if (!_copyOutMode)
+                            {
+                                // Protocol violation. CopyDone received twice.
+                                throw _connector.UnexpectedMessageReceived(msg.Code);
+                            }
+                            lock (_writeSyncObject)
+                            {
+                                _copyOutMode = false;
+                                if (_copyInMode)
+                                {
+                                    _connector.SendMessage(CopyDoneMessage.Instance);
+                                    _copyInMode = false;
+                                }
+                            }
+                            dontWait = false;
+                            continue;
+                        case BackendMessageCode.CompletedResponse:
+                            Expect<CommandCompleteMessage>(await _connector.ReadMessage(async));
+                            Expect<ReadyForQueryMessage>(await _connector.ReadMessage(async));
+                            lock (_writeSyncObject)
+                            {
+                                if (!_disposed)
+                                {
+                                    if (_connector.State == ConnectorState.Replication)
+                                        _connector.EndUserAction();
+
+                                    Cleanup();
+                                }
+                            }
+                            EndOfStream = true;
+                            if (_copyInMode || _copyOutMode)
+                            {
+                                // There is currently an issue with the PostgreSQL that we can not run a replication stream more than once in the same connection.
+                                // If this connection already rans a stream before, it should be discarded.
+                                throw new InvalidOperationException("Unexpected message received from server. If this connection already ran a replication stream before, it should be discarded. Replication connections can only run replication stream once.");
```

Comment:
This is the exception thrown when we try to replay a replication stream in the same connection. When trying to do that, we receive a CopyBoth server message, followed by `CommandComplete` and `ReadyForQuery`, without receiving `CopyData` or `CopyDone`. This behavior is unexpected and I already reported this to the PostgreSQL team.